packages/lib-auth/src/client.js

```diff
@@ -0,0 +1,320 @@
+const cookie = require('cookie')
+const isNode = require('detect-node')
+
+const createImmutable = require('./helpers/create-immutable')
+
+/**
+ * The main Client class, which provides methods to authenticate against the API and request access tokens.
+ *
+ * @class
+ */
+class Client {
+  /**
+   * Creates a new instance of the client
+   *
+   * @param {object} config - Configuration parameters for the client
+   * @param {object} httpClient - A configured HTTP client for making requests
+   */
+  constructor (config, httpClient) {
+    createImmutable(this, '_config', config)
+    createImmutable(this, '_httpClient', httpClient)
+    this._resetState()
+  }
+
+  /**
+   * Makes a request to exchange a JWT for a access token.
+   *
+   * @async
+   * @private
+   * @param {string} jwt - A JSON Web Token
+   * @returns {Promise} Resolves to the new access token.
+   */
+  async _exchangeJWTForToken (jwt) {
+    if (!jwt) {
+      return null
+    }
+
+    const data = {
+      'client_id': this._config.clientAppID,
+      'grant_type': 'password'
+    }
+
+    const config = {
+      headers: {
+        cookie: `${this._config.cookieName}=${jwt}`
+      },
+      withCredentials: true
+    }
+
+    const response = await this._httpClient.post('/oauth/token', data, config)
+    this._handleNewTokenData(response.data)
+    return this._state.accessToken
+  }
+
+  /**
+   * Gets an access token. Returns the current access token if it's still valid, otherwise attempts to try and refresh it before returning.
+   *
+   * @async
+   * @returns {Promise} Resolves to the access token, or an empty string if not available.
+   */
+  async getAccessToken() {
+    if (!this._state.accessToken) {
+      return ''
+    } else if (this._isAccessTokenValid()) {
+      return this._state.accessToken
+    } else {
+      return this._refreshTokenData()
+    }
+  }
+
+  /**
+   * Requests a cross-site request forgery (CSRF) token, which is used in the main authentication-related actions.
+   *
+   * @async
+   * @private
+   * @returns {Promise} Resolves to the CSRF token.
+   */
+  async _getCSRFToken () {
+    try {
+      const url = `/users/sign_in?now=${Date.now()}`
+      const config = { withCredentials: true }
+      const response = await this._httpClient.head(url, config)
+      const csrfToken = response.headers['x-csrf-token']
+      return csrfToken
+    } catch (error) {
+      console.error(error)
+      throw error
```

Comment:
I assume we want to examine the error in the console but not continue with authentication? I think this block should return null, or an empty string, rather than throwing.

Reply:
I chose to re-throw here to get a helpful stack trace. Returning something falsey is an option too, but you'd need to add a check in the method calling this one and throw an error there so it can be handled appropriately.

Reply:
I'm still not following why an error has to be thrown.

Reply:
The docs say this method returns a Promise that resolves to a token, so I think it will need to return something, even on error.